server/src/main/java/org/opensearch/index/translog/transfer/FileTransferTracker.java

```diff
@@ -85,38 +107,89 @@ void add(String file, boolean success) {
+        }
+    }
+
+    @Override
+    public void onFailure(TranslogCheckpointSnapshot fileSnapshot, Exception e) {
         long durationInMillis = (System.nanoTime() - fileTransferStartTime) / 1_000_000L;
         remoteTranslogTransferTracker.addUploadTimeInMillis(durationInMillis);
-        remoteTranslogTransferTracker.addUploadBytesFailed(bytesForTlogCkpFileToUpload.get(fileSnapshot.getName()));
-        add(fileSnapshot.getName(), TransferState.FAILED);
+        updateTransferStats(fileSnapshot, false);
+        addGeneration(fileSnapshot.getGeneration(), TransferState.FAILED);
     }
 
-    public void delete(List<String> names) {
+    private void updateTransferStats(TranslogCheckpointSnapshot fileSnapshot, boolean isSuccess) {
+        Long translogFileBytes = bytesForTlogCkpFileToUpload.get(fileSnapshot.getTranslogFileName());
+        if (translogFileBytes != null) {
+            if (isSuccess) {
+                remoteTranslogTransferTracker.addUploadBytesSucceeded(translogFileBytes);
+            } else {
+                remoteTranslogTransferTracker.addUploadBytesFailed(translogFileBytes);
+            }
+        }
+
+        Long checkpointFileBytes = bytesForTlogCkpFileToUpload.get(fileSnapshot.getCheckpointFileName());
+        if (checkpointFileBytes != null) {
+            if (isSuccess) {
+                remoteTranslogTransferTracker.addUploadBytesSucceeded(checkpointFileBytes);
+            } else {
+                remoteTranslogTransferTracker.addUploadBytesFailed(checkpointFileBytes);
+            }
+        }
+    }
+
+    void delete(List<String> names) {
         for (String name : names) {
             fileTransferTracker.remove(name);
         }
     }
 
+    void deleteGenerations(Set<Long> generations) {
+        for (Long generation : generations) {
+            generationTransferTracker.remove(generation);
+        }
+    }
+
     public boolean uploaded(String file) {
         return fileTransferTracker.get(file) == TransferState.SUCCESS;
     }
 
-    public Set<TransferFileSnapshot> exclusionFilter(Set<TransferFileSnapshot> original) {
+    boolean translogGenerationUploaded(Long generation) {
+        return generationTransferTracker.get(generation) == TransferState.SUCCESS;
+    }
+
+    Set<TranslogCheckpointSnapshot> exclusionFilter(Set<TranslogCheckpointSnapshot> original) {
         return original.stream()
-            .filter(fileSnapshot -> fileTransferTracker.get(fileSnapshot.getName()) != TransferState.SUCCESS)
+            .filter(fileSnapshot -> generationTransferTracker.get(fileSnapshot.getGeneration()) != TransferState.SUCCESS)
             .collect(Collectors.toSet());
     }
 
+    Set<Long> allUploadedGeneration() {
+        Set<Long> successGenFileTransferTracker = new HashSet<>();
+        generationTransferTracker.forEach((k, v) -> {
+            if (v == TransferState.SUCCESS) {
+                successGenFileTransferTracker.add(k);
+            }
+        });
```

Comment:
this is also duplicated twice, can we have a common method that can be reused by both method.